core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/interceptor/RequestCompressionInterceptor.java

```diff
@@ -0,0 +1,292 @@
+ * Compression of payload will be completed prior to checksum calculation. The Content-Encoding header will be updated
+ * accordingly.
+ */
+@SdkInternalApi
+public class RequestCompressionInterceptor implements ExecutionInterceptor {
+
+    private static final int DEFAULT_MIN_COMPRESSION_SIZE = 10_240;
+    private static final int MIN_COMPRESSION_SIZE_LIMIT = 10_485_760;
+    private static final Supplier<ProfileFile> PROFILE_FILE = ProfileFile::defaultProfileFile;
+    private static final String PROFILE_NAME = ProfileFileSystemSetting.AWS_PROFILE.getStringValueOrThrow();
+
+    @Override
+    public SdkHttpRequest modifyHttpRequest(Context.ModifyHttpRequest context, ExecutionAttributes executionAttributes) {
+        if (!shouldCompress(context, executionAttributes)) {
+            return context.httpRequest();
+        }
+
+        Compressor compressor = resolveCompressionType(executionAttributes);
+        if (executionAttributes.getAttribute(SdkInternalExecutionAttribute.REQUEST_COMPRESSION).isStreaming()) {
+            return updateContentEncodingHeader(context.httpRequest(), compressor);
+        }
+
+        SdkHttpFullRequest sdkHttpFullRequest = (SdkHttpFullRequest) context.httpRequest();
+        InputStream inputStream = sdkHttpFullRequest.contentStreamProvider().get().newStream();
+        InputStream compressedStream = compressor.compress(inputStream);
+        SdkHttpRequest sdkHttpRequest =
+            sdkHttpFullRequest.toBuilder()
+                              .contentStreamProvider(() -> compressedStream)
+                              .build();
+        sdkHttpRequest = updateContentEncodingHeader(sdkHttpRequest, compressor);
+        return updateContentLengthHeader(sdkHttpRequest);
+    }
+
+    @Override
+    public Optional<RequestBody> modifyHttpContent(Context.ModifyHttpRequest context, ExecutionAttributes executionAttributes) {
+        if (!context.requestBody().isPresent() || !shouldCompress(context, executionAttributes)) {
+            return context.requestBody();
+        }
+
+        Compressor compressor = resolveCompressionType(executionAttributes);
+        RequestBody requestBody = context.requestBody().get();
+
+        if (isTransferEncodingChunked(context)) {
+            InputStream compressedStream = compressor.compress(requestBody.contentStreamProvider().newStream());
+            try {
+                byte[] compressedBytes = IoUtils.toByteArray(compressedStream);
+                return Optional.of(RequestBody.fromBytes(compressedBytes));
+            } catch(IOException e){
+                throw SdkClientException.create(e.getMessage(), e);
+            }
+        }
+
+        CompressionContentStreamProvider streamProvider =
+            new CompressionContentStreamProvider(requestBody.contentStreamProvider(), compressor);
+        return Optional.of(RequestBody.fromContentProvider(streamProvider, requestBody.contentType()));
```

Comment:
Why is there separate logic for chunked encoding? Does this interceptor need to care about chunked encoding at all?